src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java

```diff
@@ -0,0 +1,661 @@
+    if ((column.getScale() != null && column.getScale() != 0)
+        || physicalType == ColumnPhysicalType.SB16) {
+      parquetType =
+          Types.primitive(PrimitiveType.PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY, repetition)
+              .length(16)
+              .as(decimalLogicalTypeAnnotation)
+              .id(id)
+              .named(name);
+    } else {
+      switch (physicalType) {
+        case SB1:
+        case SB2:
+        case SB4:
+          parquetType =
+              Types.primitive(PrimitiveType.PrimitiveTypeName.INT32, repetition)
+                  .as(decimalLogicalTypeAnnotation)
+                  .id(id)
+                  .length(4)
+                  .named(name);
+          break;
+        case SB8:
+          parquetType =
+              Types.primitive(PrimitiveType.PrimitiveTypeName.INT64, repetition)
+                  .as(decimalLogicalTypeAnnotation)
+                  .id(id)
+                  .length(8)
+                  .named(name);
+          break;
+        default:
+          throw new SFException(
+              ErrorCode.UNKNOWN_DATA_TYPE, column.getLogicalType(), column.getPhysicalType());
+      }
+    }
+    return parquetType;
+  }
+
+  /**
+   * Get the parquet type for column of a Snowflake time logical type.
+   *
+   * @param scale column scale
+   * @param physicalType Snowflake physical type of column
+   * @param logicalType Snowflake logical type of column
+   * @param supportedPhysicalTypes supported Snowflake physical types for the given column
+   * @param repetition parquet repetition type of column
+   * @param id column id in Snowflake table schema
+   * @param name column name
+   * @return column parquet type
+   */
+  private static Type getTimeColumnParquetType(
+      Integer scale,
+      ColumnPhysicalType physicalType,
+      ColumnLogicalType logicalType,
+      Set<ColumnPhysicalType> supportedPhysicalTypes,
+      Type.Repetition repetition,
+      int id,
+      String name) {
+    if (scale == null || scale > 9 || scale < 0 || !supportedPhysicalTypes.contains(physicalType)) {
+      throw new SFException(
+          ErrorCode.UNKNOWN_DATA_TYPE,
+          "Data type: " + logicalType + ", " + physicalType + ", scale: " + scale);
+    }
+    LogicalTypeAnnotation.TimeUnit timeUnit = getTimeUnitFromScale(scale);
+
+    PrimitiveType.PrimitiveTypeName type = getTimePrimitiveType(physicalType);
+    if (physicalType == ColumnPhysicalType.SB16) {
+      LogicalTypeAnnotation typeAnnotation = LogicalTypeAnnotation.decimalType(scale, 38);
+      return Types.primitive(type, repetition).as(typeAnnotation).length(16).id(id).named(name);
+    } else {
+      LogicalTypeAnnotation typeAnnotation =
+          logicalType == ColumnLogicalType.TIME
+              ? LogicalTypeAnnotation.timeType(false, timeUnit)
+              : LogicalTypeAnnotation.timestampType(false, timeUnit);
+      return Types.primitive(type, repetition).as(typeAnnotation).id(id).named(name);
```

Comment:
are we missing `.length()`?

Reply:
I see this is missing in some other places as well. Is the length not supposed to be defined for every `Types.primitive()`?

Reply:
This is my understanding, the bit/byte length of primitives is predefined (int64 etc). I think it even throws because length cannot be changed for primitives.

Reply:
But we have in `L#230`.length(4) for `PrimitiveType.PrimitiveTypeName.INT32` and `L#241` .length(8). The reason I am asking is because I noticed that in tests for time columns the `type length` is 0.